### PR TITLE
feat : Fundraising campaign 

### PR DESCRIPTION
**Pull request by saswat10**
**What kind of change does this PR introduce?**
Feature

**Issue Number:** -->  #1663 Add Support of Fundraising Campaigns

**Did you add tests for your changes?** --> Yes

**Summary**
This Pull Request introduces the addition of *fundraising campaigns* schema and the required CRUD operations , along with the test files.

**Does this PR introduce a breaking change?** --> No

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?** --> Yes

### CODE REVIEW

**Comment by palisadoes** [comment]
Please fix the conflicting files

**Comment by sumitra19jha** [comment]
Please fix the conflict for review

**Comment by saswat10** [comment]
> Please fix the conflict for review

Do I need to fix it? Cause it shows I don't have the write access

**Comment by saswat10** [comment]
Please can you look at the Pull Request, I have resolved the conflicts in the files.

**Review by palisadoes — CHANGES_REQUESTED** [review]
@saswat10 

1. Restore .husky/post-merge
2. Restore .husky/pre-commit

They are an integral part of our code quality process. You will need to make sure your code complies with our linting standards.

1. Do this immediately. We will not assign reviewers until this is done.
2. This is why the linting tests are failing in the PR
3. It is also why the authorized access test and application tests are failing

**Comment by saswat10** [comment]
@palisadoes  please check now, I have made necessary changes.

**Comment by palisadoes** [comment]
1. You need to make sure all your linting tests pass or else the PR tests will never pass.
2. Then when the linting passes, you'll need to make sure the other tests pass and that the code coverage doesn't decline. We are at close to 97% testing coverage and we cannot afford to make that drop below our 95% target. The code was really buggy in the past and the tests have made it significantly better.
3. Make sure to test before submitting further updates and verify the test code coverage in your local repo. The CONTRIBUTING.md file goes over this.

It's difficult for us to be constantly approving the workflows for every minor commit. 

Please take the time to have better testing and linting before the next commit.

**Comment by palisadoes** [comment]
FYI. Fix these first

image

**Comment by palisadoes** [comment]
@saswat10 How is progress on this?

**Comment by palisadoes** [comment]
We need to have this implemented soon. See the comments above

**Comment by saswat10** [comment]
> 1. How is progress on this?

 Currently working on it.

> 2. Why has the definition for the Fund not been submitted?

Will raise a PR by tomorrow.

> 3. Why doesn't the FundCampaign refer to a Fund in the Schema?

Since you said that fund campaign cannot have more than one associated funds. But will I will add a Fund schema which would refer to the parent fund  in the fund campaign. 
image

**Comment by palisadoes** [comment]
We agreed on your proposed schema.

```
Funds can have many associated campaigns and not vice versa
```

ie. Campaigns can only be associated with one fund.

It would be good to see the end result of your work. We have a few follow up issues that depend on this one being completed.

**Comment by palisadoes** [comment]
Where is the fund in the schema? This does not comply with the issue requirements.

**Comment by saswat10** [comment]
> Where is the fund in the schema? This does not comply with the issue requirements.
 
I haven't  tested that code yet, It is in a different branch , I will add by today positively.

**Comment by palisadoes** [comment]
Why was this closed? We have spent so much time reviewing you work.

**Comment by saswat10** [comment]
> Why was this closed? We have spent so much time reviewing you work.

I was checking for the workflow
Making minor commits can send message to the reviewer, hence I closed the issue


I have reopened the issue, but made it into a draft

**Comment by palisadoes** [comment]
1. We don't accept draft PRs.
1. We get notifications with every commit and comment no matter the type of PR.

Convert to a regular PR or close this and practice on your forked repository

**Comment by palisadoes** [comment]
You can test on your local repo. We have to separately approve each of your commits

**Comment by palisadoes** [comment]
Let us know when you are ready

**Comment by palisadoes** [comment]
We are not going to review a draft PR

**Comment by palisadoes** [comment]
@sumitra19jha @xoldyckk @EshaanAgg @JamaicanFriedChicken 

This is ready for review

**Comment by palisadoes** [comment]
Please fix the failing tests. Our test code coverage system will fail if any of these conditions occur:

1) The overall code coverage drops below the target threshold of the repository
2) Any file in the pull request has code coverage levels below the repository threshold

We do this to ensure the overall reliability of the code base is maintained. This is a prerequisite for getting your PR approved. Assigned reviewers regularly review the PR queue and tend to focus on PRs that are passing.

**Comment by saswat10** [comment]
> Please fix the failing tests. Our test code coverage system will fail if any of these conditions occur:
> 
> 1. The overall code coverage drops below the target threshold of the repository
> 2. Any file in the pull request has code coverage levels below the repository threshold
> 
> We do this to ensure the overall reliability of the code base is maintained. This is a prerequisite for getting your PR approved. Assigned reviewers regularly review the PR queue and tend to focus on PRs that are passing.

Ok. Working on it

**Comment by palisadoes** [comment]
Please address the decline in code test coverage.

**Comment by saswat10** [comment]
@palisadoes I'm facing a problem, the autogenerated schema for the fund schema is bit weird like

image
 
Due to this whatever test case I'm writing for the FundCampaign helper functions, it is showing errors....

Why can't I get something simpler auto generated schema like this

image
image

**Comment by saswat10** [comment]
> @palisadoes I'm facing a problem, the autogenerated schema for the fund and fund campaign schemas is bit weird like
> 
> image
> 
> Due to this whatever **test case I'm writing for the FundCampaign helper functions, it is showing errors....**
> 
> I not getting something simpler auto generated schema like this
> 
> image image

Due to this I had skipped writing test case for the Fund Campaign helper functions like creator.ts, fund.ts and pledge.ts

File: schema.graphql
Comment (EshaanAgg) [inline schema.graphql:65]: You have not created the type definitions properly, which is leading to the items being nullable and thus, the type errors you are getting. Refer to the comments.

File: src/models/CampaignPledge.ts
Comment (EshaanAgg) [inline src/models/CampaignPledge.ts:14]: How many members and campaigns is a CampaignPledge supposed to hold? Will it be a small number like 10-15, or can it be something infinitely scalable (that is something in the order of let's say 5% of members of the organization)? If it's the later, we should be using a joining table and not the direct embedding of ducments.

File: src/models/FundCampaign.ts
Comment (EshaanAgg) [inline src/models/FundCampaign.ts:22]: Many fields are repeated between the models `Fund`, `FundCampaign` and `CampaignPlegde`. The relationship should be modelled through parent and children, and only the id of the parent should be stored on the children to reduce data redundancy.

File: src/typeDefs/inputs.ts
Comment (EshaanAgg) [inline src/typeDefs/inputs.ts:127]: Can these be `null`? Should the FundCampaign always be associated with a pledge or fund?

File: src/typeDefs/mutations.ts
Comment (EshaanAgg) [inline src/typeDefs/mutations.ts:27]:
```suggestion
    addParentFundtoCampaign(data: addParentFund!): FundCampaign! @auth
```

File: src/typeDefs/mutations.ts
Comment (EshaanAgg) [inline src/typeDefs/mutations.ts:29]:
```suggestion
    addPledgeToCampaign(data: addPledge!): FundCampaign! @auth
```

File: src/typeDefs/mutations.ts
Comment (EshaanAgg) [inline src/typeDefs/mutations.ts:78]: Remove this if this is a comment.

File: src/typeDefs/mutations.ts
Comment (EshaanAgg) [inline src/typeDefs/mutations.ts:219]:
```suggestion
    updateFundCampaign(id: ID!, data: UpdateFundCampaignInput!): FundCampaign!
```

File: src/typeDefs/queries.ts
Comment (EshaanAgg) [inline src/typeDefs/queries.ts:42]: No need for this. These should be implemented as a field level resolver either on the `Organization` model, or the other relevant parent models to `FundCampaign`

File: src/typeDefs/types.ts
Comment (EshaanAgg) [inline src/typeDefs/types.ts:96]:
```suggestion
    creator: User!
```

File: src/typeDefs/types.ts
Comment (EshaanAgg) [inline src/typeDefs/types.ts:97]:
```suggestion
    organization: Organization!
```

File: src/typeDefs/types.ts
Comment (EshaanAgg) [inline src/typeDefs/types.ts:102]:
```suggestion
    campaigns: [FundCampaign!]!
```

File: src/typeDefs/types.ts
Comment (EshaanAgg) [inline src/typeDefs/types.ts:109]:
```suggestion
    creator: User!
```

File: src/typeDefs/types.ts
Comment (EshaanAgg) [inline src/typeDefs/types.ts:117]:
```suggestion
    parentFund: Fund!
```

File: src/typeDefs/types.ts
Comment (EshaanAgg) [inline src/typeDefs/types.ts:118]:
```suggestion
    pledge: CampaignPledge!
```

File: src/typeDefs/types.ts
Comment (EshaanAgg) [inline src/typeDefs/types.ts:123]:
```suggestion
    creator: User!
```

File: src/typeDefs/types.ts
Comment (EshaanAgg) [inline src/typeDefs/types.ts:126]:
```suggestion
    members: [User!]!
```

File: src/typeDefs/types.ts
Comment (EshaanAgg) [inline src/typeDefs/types.ts:127]:
```suggestion
    campaigns: [FundCampaign!]!
```

File: src/typeDefs/types.ts
Comment (EshaanAgg) [inline src/typeDefs/types.ts:441]:
```suggestion
    fundCampaigns: [FundCampaign!]!
```

**Review by EshaanAgg — CHANGES_REQUESTED** [review]
I am not very sure about the organization of the three models and how their hierarchy would be, so I may have added some not-null assertions where they may be nullable, so please take this review with caution. But the fact remains that you need to update the type definitions as mentioned here. 

Also, please try to make smaller PRs. Ideally, you should only add the schema definitions and typedefs in a PR, and open a PR for every 2-3 resolvers. It is not possible to holistically review a PR of this size. Please keep this in mind for the next time.

**Comment by palisadoes** [comment]
Please fix the conflicting files and address the reviewer requests

**Comment by palisadoes** [comment]
When do you think this will be ready for review?

**Comment by sumitra19jha** [comment]
Please resolve conflict

**Comment by palisadoes** [comment]
*NOTE* Read very carefully

1. We just merged this PR which upgraded the prettier package. 
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1840
1. It reformatted over 400 files. This will put your PR at risk of extensive merge conflicts.    
1. I suggest that before you merge your code to the latest upstream, or manually fix these merge conflicts that you:
    1. upgrade your `prettier` in your local branch to the same version
    1. run `prettier`
    
This will help to reduce the number of future merge conflicts for your PR.

**Comment by palisadoes** [comment]
Closing due to inactivity and the fact that this has been open for a month.

**Comment by varshith257** [comment]
@palisadoes  Waiting for your update on this issue assignment. Please assign me this issue. I am eagerly waiting to contribute to this

**Comment by varshith257** [comment]
@palisadoes Waiting for the update on the assignment of the issue for quite a long time

**Pull request closed without merging**
